### PR TITLE
Query ontology database for experiment creation/editing

### PR DESCRIPTION
**Pull request by wow-such-code**
* Provides methods for lazy querying of the ontology_classes table with high performance
* This uses a fulltext index on the label column, the MyISAM engine (InnoDB asterisk matching is currently bugged) with min word length 2 (to find the shortest species, e.g. _Ia Io_)
* To provide search results for shorter (or empty) filter strings, two other queries are used
* Connects dropdowns for species, specimen, analyte to this table (a list of ontologies to use can be provided, e.g. "NCBITaxon" for Species), creates boolean mode matching terms (e.g. +Homo +sa*) and queries the database accordingly
* Ontology terms are mapped to Species, Specimen, Analyte upon display in the dropdowns. This might need to be changed later to fetch more information than the label
* Changes openBIS sample registration to set default analyte if it cannot be found in the openBIS vocabulary. The respective ontology term will now be stored in a new free text property instead
* Provides some other classes that can be extended to do more with ontologies in the future

### CODE REVIEW

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/update/ExperimentUpdateDialog.java [hunk 1]
```diff
@@ -9,18 +9,27 @@
...
 import java.util.HashSet;
 import java.util.List;
 import java.util.Objects;
 import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
 import life.qbic.datamanager.views.events.UserCancelEvent;
 import life.qbic.datamanager.views.general.DialogWindow;
+import life.qbic.datamanager.views.projects.project.experiments.experiment.create.ExperimentAddDialog;
+import life.qbic.datamanager.views.projects.project.experiments.experiment.create.ExperimentAddDialog.ExperimentDraft;
```
Comment (KochTobi) [inline user-interface/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/update/ExperimentUpdateDialog.java:28]: Should not use the same data object. We introduce a dependency here where none is required. The dialogs may change for different reasons at different times.

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/update/ExperimentUpdateDialog.java [hunk 2]
```diff
@@ -185,88 +205,4 @@ public Optional<ExperimentDraft> getOldDraft() {
...
-
-    @Override
-    public int hashCode() {
-      int result = experimentName != null ? experimentName.hashCode() : 0;
-      result = 31 * result + species.hashCode();
-      result = 31 * result + specimen.hashCode();
-      result = 31 * result + analytes.hashCode();
-      return result;
-    }
-  }
```
Comment (KochTobi) [inline user-interface/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/update/ExperimentUpdateDialog.java:271]: Why did you remove this ExperimentDraft object?

**Review by Steffengreiner — CHANGES_REQUESTED** [review]
Heya @sven1103 and @wow-such-code! Nice work the Ontologies look impressive and load really fast! 🚤
However i found some inconsistencies which should be addressed

File: project-management/src/main/java/life/qbic/projectmanagement/application/OntologyClassEntity.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    this.ontologyVersion = ontologyVersion;
+  }
+
+  public String getOntologyIri() {
+    return ontologyIri;
+  }
+
+  public void setOntologyIri(String ontologyIri) {
+    this.ontologyIri = ontologyIri;
+  }
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management/src/main/java/life/qbic/projectmanagement/application/OntologyClassEntity.java:67]: this class was provided by @sven1103 , I'll leave that for him to answer.

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/ontology/OntologyTermRepository.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ * Simple repository to query concise ontology term information
+ *
+ * @since 1.0.0
+ */
+public interface OntologyTermRepository extends
+    PagingAndSortingRepository<OntologyClassEntity, Long> {
+
+  @Query(value = "SELECT * FROM ontology_classes WHERE MATCH(label) AGAINST(?1 IN BOOLEAN MODE) AND ontology in (?2) ORDER BY length(label);",
+      countQuery = "SELECT count(*) FROM ontology_classes WHERE MATCH(label) AGAINST(?1 IN BOOLEAN MODE) AND ontology in (?2);",
+      nativeQuery = true)
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/ontology/OntologyTermRepository.java:20]: WHERE MATCH(label) AGAINST queries are not handled automatically, from what I can tell.

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/update/ExperimentUpdateDialog.java (hunk 2 above)
Comment (wow-such-code) [reply user-interface/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/update/ExperimentUpdateDialog.java]: persuaded, I reintroduced it

File: user-interface/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/create/ExperimentAddDialog.java [hunk 5]
```diff
@@ -40,56 +45,51 @@ public class ExperimentAddDialog extends DialogWindow {
...
     Span experimentDescription = new Span(
         "Please specify the sample origin information of the samples. Multiple "
             + "values are allowed!");
 
     MultiSelectComboBox<Species> speciesBox = new MultiSelectComboBox<>("Species");
-    speciesBox.setRequired(true);
-    speciesBox.addClassNames(CHIP_BADGE, WIDTH_INPUT);
+    initComboBoxWithDatasource(speciesBox, List.of("NCBITaxon"),
+        term -> new Species(term.getLabel()));
+    speciesBox.setItemLabelGenerator(Species::label);
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply user-interface/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/create/ExperimentAddDialog.java]: solved

File: project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java [hunk 6]
```diff
@@ -72,6 +72,7 @@ public class OpenbisConnector implements ExperimentalDesignVocabularyRepository,
   private static final String DEFAULT_SPACE_CODE = "DATA_MANAGER_SPACE";
   private static final String DEFAULT_SAMPLE_TYPE = "Q_TEST_SAMPLE";
   private static final String DEFAULT_EXPERIMENT_TYPE = "Q_SAMPLE_PREPARATION";
+  private static final String DEFAULT_ANALYTE_TYPE = "OTHER";
   private static final String DEFAULT_DELETION_REASON = "Commanded by data manager app";
 
   private final AnalyteTermMapper analyteMapper = new SimpleOpenBisTermMapper();
```
[the remark this answers is not in the document]
Comment (wow-such-code) [reply project-management-infrastructure/src/main/java/life/qbic/projectmanagement/infrastructure/sample/openbis/OpenbisConnector.java:78]: good point, I will hunt them down

File: project-management/src/main/java/life/qbic/projectmanagement/application/OntologyClassEntity.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sven1103) [reply project-management/src/main/java/life/qbic/projectmanagement/application/OntologyClassEntity.java:67]: Java does not yet support IRIs.

**Review by Steffengreiner — APPROVED** [review]
LGTM 🚀
